infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java

```
@@ -0,0 +1,213 @@
+                return;
+            }
+
+            // Store clientID as a channel attribute.
+            AttributeKey<UUID> clientIdAttrKey = AttributeKey.valueOf("ClientID");
+            ctx.channel().attr(clientIdAttrKey).set(clientId);
+            log.info("Handshake validated by Server.");
+            log.debug("Sending handshake response: Node Id: " + this.nodeId +
+                    " Corfu Version: " + this.corfuVersion);
+
+            CorfuMsg handshakeResponse = CorfuMsgType.HANDSHAKE_RESPONSE
+                    .payloadMsg(new HandshakeResponse(this.nodeId, this.corfuVersion));
+            ctx.writeAndFlush(handshakeResponse);
+
+            // Flush messages in queue
+            log.debug("There are {" + this.messages.size() + "} messages in queue to " +
+                    "be flushed.");
+            for (CorfuMsg message : this.messages) {
+                ctx.writeAndFlush(message);
+            }
+
+            // Remove this handler from the pipeline; handshake is completed.
+            log.info("Removing handshake handler from pipeline.");
+            ctx.pipeline().remove(this);
+            this.fireHandshakeSucceeded();
+        }
+    }
+
+
+    @Override
+    public void channelActive(final ChannelHandlerContext ctx) throws Exception {
+        this.group.add(ctx.channel());
+        log.info("Incoming connection established from: " +
+                ctx.channel().remoteAddress());
+
+        // Start timeout checker. Wait for handshake to complete, otherwise close connection.
+        new Thread() {
+
+            @Override
+            public void run() {
+                try {
+                    log.debug("Wait for the handshake to complete, then disconnect");
+                    latch.await(timeoutInMillis, TimeUnit.MILLISECONDS);
+                } catch (InterruptedException e) {
+                    log.error("Handshake timeout checker: interrupted!");
+                    e.printStackTrace();
```

Should re-throw the exception :).